client-spark/spark-2/src/main/java/org/apache/spark/shuffle/celeborn/HashBasedShuffleWriter.java

```diff
@@ -319,6 +324,21 @@ private void flushSendBuffer(int partitionId, byte[] buffer, int size)
     writeMetrics.incWriteTime(System.nanoTime() - start);
   }
 
+  private void abort(IOException lastIOException) throws IOException {
```

Comment:
QQ: Looking at this more, `abort` is called only when some exception was thrown - right ?
If yes, we can simply make this:

```
 public void write(scala.collection.Iterator<Product2<K, V>> records) throws IOException {
    boolean needAbort = true;
    try {
      // <current logic>
      close();
      needAbort = false;
    } catch (InterruptedException e) {
      TaskInterruptedHelper.throwTaskKillException();
    } finally {
      if (needAbort) {
        abort(false);
      }
    }

<snip>

private void abort(boolean throwTaskKilledOnInterruption) {
    try {
      dataPusher.waitOnTermination();
      sendBufferPool.returnPushTaskQueue(dataPusher.getIdleQueue());
    } catch (InterruptedException e) {
       if (throwTaskKilledOnInterruption) {
         TaskInterruptedHelper.throwTaskKillException();
       }
    }
}

```
I am keeping `throwTaskKilledOnInterruption` explicitly here to illustrate - but it should always be `false` - since there will always be an exception raised when needAbort == true - right ?

Reply:
Thank you for your patient guidance. I may not have understood the problem you mentioned at first. I also fixed this problem in `SortBasedShuffleWriter#abort`.